### DISCUSSION TITLE
@Modifying은 무엇이고 주의할 점은 무엇일까요?

### DISCUSSION THREAD

**Discussion by le2sky** [2024-woowacourse-study/level-interview#202] [Spring Data]
뭘까용

**Comment by alstn113** [comment]
`@Modifying`은 은 Spring Data JPA에서 사용되는 애노테이션으로, @Query 어노테이션(JPQL Query, Native Query)을 통해 작성된 INSERT, UPDATE, DELETE (SELECT 제외) 쿼리에서 사용된다.
clearAutomatically와 flushAutomatically 속성이 있으며, 주로 벌크 연산(여러 레코드 수정)을 수행할 때 사용한다.
JPA Entity 생명주기를 무시하고 쿼리가 실행되기 때문에 영속성 컨텍스트 관리에 주의해야 한다.

### 벌크 연산
벌크 연산이란, 단건 UPDATE, DELETE 연산을 제외한 다건의 UPDATE, DELETE 연산을 하나의 쿼리로 하는 것을 의미한다. 
JPA에서 단건 UPDATE 같은 경우에는 Dirty Checking(변경 감지)를 통해서 수행되거나 save()를 통해 수행할 수 있다. 
DELETE 경우 단건, 다건 쿼리 메서드로 제공된다.
@Modifying과 @Query를 사용해서 벌크 연산을 할 수 있는데 이 방법의 장점은 JPQL을 자유롭게 정의하여 사용할 수 있고, 
하나의 쿼리로 많은 데이터를 변경할 수 있다는 점이다.
참고로 @Query에 벌크 연산 쿼리를 작성하고, @Modifying 어노테이션을 붙이지 않으면 InvalidDataAccessApiUsageException이 발생한다.

### clearAutomatically
clearAutomatically 속성은 벌크 연산 후 영속성 컨텍스트를 자동으로 clear할지 여부를 결정한다. 기본값은 false이다.
벌크 연산은 DB에 직접 접근하기 때문에 영속성 컨텍스트의 상태를 무시하고 쿼리를 실행한다.
1차 캐시에 엔티티를 저장하고, 벌크 연산을 실행한 후에는 1차 캐시와 DB의 데이터가 일치하지 않을 수 있다.
이 때 조회를 하는 경우 1차 캐시에 저장된 데이터를 우선적으로 사용하기 때문에 데이터 불일치 문제가 발생할 수 있다.
clearAutomatically를 true로 설정하면 수정 쿼리 실행 후 영속성 컨텍스트를 자동으로 clear하여 1차 캐시를 비우고, 새로운 데이터를 DB에서 다시 조회하여 일관성을 유지할 수 있다.

### flushAutomatically
flushAutomatically 속성은 벌크 연산 전에 영속성 컨텍스트를 flush할지 여부를 결정한다. 기본값은 false이다.
하지만 벌크 연산을 수행하기 전에 flush를 하는데 그 이유는 Hibernate의 FlushModeType가 default로 AUTO로 설정되어 있기 때문이다.
FlushModeType.AUTO는 쿼리 실행 전에 영속성 컨텍스트를 Flush한다.

참고: https://frogand.tistory.com/174

**Comment by 3Juhwan** [comment]
`@Modifying`은 Spring Data JPA에서 데이터베이스의 데이터를 수정하는 쿼리를 정의할 때 사용하는 어노테이션입니다. 이 어노테이션은 데이터 삽입, 업데이트, 삭제 등의 작업을 수행하는 메서드에 적용됩니다. 반드시 `@Query` 어노테이션과 함께 사용되어야 하고 트랜잭션 범위 안에 있어야 합니다. 

`@Modifying`을 사용할 때 주의할 점은 다음과 같습니다. 

**읽기 전용 트랜잭션에서 사용 금지**

`@Modifying`이 적용된 메서드는 기본적으로 트랜잭션 내에서 실행되어야 합니다. Spring Data JPA는 `@Transactional` 어노테이션을 사용하여 트랜잭션을 관리합니다. 따라서 `@Modifying` 어노테이션이 있는 메서드에는 `@Transactional` 어노테이션을 추가하는 것이 좋습니다.

**벌크 업데이트/삭제**

벌크 연산은 영속성 컨텍스트를 거치지 않아 변경 감지와 같은 처리를 생략합니다. 이 때문에 벌크 업데이트나 삭제를 수행할 때는 영속성 컨텍스트와 데이터베이스 동기화 문제가 발생할 수 있습니다. 가령, 벌크 연산으로 변경된 데이터가 영속성 컨텍스트에 남아 있어 조회된 데이터와 데이터베이스 사이의 일관성이 깨집니다. 이런 경우에는 엔티티 매니저를 수동으로 플러시하거나 클리어하는 방법을 사용할 수 있습니다.

**반환 타입**

반환 타입은 void 또는 int/Integer이어야 합니다. 반환되는 값은 `@Modifying` 쿼리 연산을 통해 변경된 레코드 수입니다. 

</br>

`@Modifying`의 옵션은 다음과 같습니다. 옵션을 통해 영속성 컨텍스트와 데이터베이스 간의 일관성을 유지할 수 있습니다. 

- flushAutomatically: 변경 쿼리가 나가기 전에 영속성 컨텍스트를 flush합니다. 
- clearAutomatically: 변경 쿼리를 실행한 후에 영속성 컨텍스트를 clear합니다.
